sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java

```diff
@@ -431,21 +431,28 @@ private void applySessionToken(RxDocumentServiceRequest request) {
         Map<String, String> headers = request.getHeaders();
         Objects.requireNonNull(headers, "RxDocumentServiceRequest::headers is required and cannot be null");
 
+        String requestConsistencyLevel = headers.get(HttpConstants.HttpHeaders.CONSISTENCY_LEVEL);
+
+        boolean sessionTokenApplicable =
+            Strings.areEqual(requestConsistencyLevel, ConsistencyLevel.SESSION.toString()) ||
+                (this.defaultConsistencyLevel == ConsistencyLevel.SESSION &&
+                    // skip applying the session token when Eventual Consistency is explicitly requested
+                    // on request-level for data plane operations.
+                    // The session token is ignored on teh backend/gateway in this case anyway
+                    // and the session token can be rather large (even run in the 16 KB header length problem
+                    // on the gateway - so not worth sending when not needed
+                    (!request.isReadOnlyRequest() ||
+                        request.getResourceType() != ResourceType.Document ||
```

Comment:
1. Should execute stored procedure be included?
2. What about get query plan? That never needs the session token.

Reply:
Query plan - good point. Stored Procedure I would not want to touch right now. Semantics aren't clear enough for me - and I am not aware of a single scenario where customers would want to adjust consistency level for SPs individually.

Reply:
Fixed in next iteration